cloudify_gcp/compute/instance.py

```diff
@@ -277,136 +304,191 @@ def get_instance_ssh_keys():
+
+    utils.delete_if_not_external(instance)
 
 
 @operation
 @utils.throw_cloudify_exceptions
-def remove_instance_tag(instance_name, tag, **kwargs):
-    gcp_config = utils.get_gcp_config()
-    if not instance_name:
-        return
-    gcp_config['network'] = utils.get_gcp_resource_name(gcp_config['network'])
-    instance = Instance(gcp_config,
-                        ctx.logger,
-                        name=instance_name)
-    instance.remove_tags([utils.get_gcp_resource_name(t) for t in tag])
+def add_instance_tag(instance_name, zone, tag, **kwargs):
+    if tag:
+        config = utils.get_gcp_config()
+        config['network'] = utils.get_gcp_resource_name(config['network'])
+        instance = Instance(config,
+                            ctx.logger,
+                            name=instance_name,
+                            zone=zone,
+                            )
+        instance.set_tags([utils.get_gcp_resource_name(t) for t in tag])
 
 
 @operation
 @utils.throw_cloudify_exceptions
-def add_external_ip(instance_name, **kwargs):
+def remove_instance_tag(instance_name, zone, tag, **kwargs):
+    config = utils.get_gcp_config()
+    if instance_name:
+        config['network'] = utils.get_gcp_resource_name(config['network'])
+        instance = Instance(config,
+                            ctx.logger,
+                            name=instance_name,
+                            zone=zone,
+                            )
+        instance.remove_tags([utils.get_gcp_resource_name(t) for t in tag])
+
+
+@operation
+@utils.throw_cloudify_exceptions
+def add_external_ip(instance_name, zone, **kwargs):
     gcp_config = utils.get_gcp_config()
     # check if the instance has no external ips, only one is supported so far
     gcp_config['network'] = utils.get_gcp_resource_name(gcp_config['network'])
     ip_node = ctx.target.node
-    instance = Instance(gcp_config,
-                        ctx.logger,
-                        name=instance_name)
+
+    # Might be overridden by either `use_external_resource` or a connected
+    # Address
+    ip_address = ''
+
+    instance = Instance(
+            gcp_config,
+            ctx.logger,
+            name=instance_name,
+            zone=zone,
+            )
+
     if ip_node.properties[constants.USE_EXTERNAL_RESOURCE]:
-        ip_address = ip_node.properties['ip_address']
+        ip_address = (
+                ip_node.properties.get('ip_address') or
+                ctx.target.instance.runtime_properties.get(constants.IP)
+                )
         if not ip_address:
             raise GCPError('{} is set, but ip_address is not set'
                            .format(constants.USE_EXTERNAL_RESOURCE))
-        instance.add_access_config(ip_address)
-    else:
-        instance.add_access_config()
+    elif ip_node.type == 'cloudify.gcp.nodes.Address':
+        ip_address = ctx.target.instance.runtime_properties['address']
+
+    instance.add_access_config(ip_address)
```

Comment:
Good point. We should be raising a `NonRecoverableError` if we get here and `ip_address` is still `''` (will probably fix with an `else:` above)